downsample_cells.R

```diff
@@ -66,68 +59,82 @@ option_list = list(
 )
 
 opt = wsc_parse_args(option_list, mandatory = c('expression_data', 'metadata', 'output_dir', 'metadata_upd'))
-suppressPackageStartupMessages(require(Matrix))
+
+if (is.na(opt$expression_data) || ! dir.exists(opt$expression_data)){
+  stop("Provide a valid directory for --expression-data")
+}else if (is.na(opt$metadata) || ! file.exists(opt$metadata)){
+  stop("Provide a valid cell metadata file")
+}
+
+# Don't parse whole matrix up-front, we might not need to do anything...
 
 expr_data = opt$expression_data
 genes = read.csv(paste(expr_data, "genes.tsv", sep="/"), sep="\t", stringsAsFactors = FALSE, header = FALSE)
 barcodes = read.csv(paste(expr_data, "barcodes.tsv", sep="/"), sep="\t", stringsAsFactors = FALSE, header = FALSE)
 cell_num_limit = floor(opt$array_size_limit / nrow(genes))
 current_cell_num = nrow(barcodes)
 
-# if no down-samling is required, simply return the output
+# if no down-samling is required return a special status code to let the user know
 if(current_cell_num <= cell_num_limit){
-    system(paste("mv", opt$expression_data, opt$output_dir, sep=" "))
-    system(paste("mv", opt$metadata, opt$metadata_upd, sep=" "))
-    cat(paste("Matrix not large enough, down-sampling is not required.\nData are moved to ",
-               opt$output_dir, " and ", opt$metadata_upd, "\n", sep=""))
-    quit(status = 0)
+  write("No downsampling required", stderr())
+  quit(status = 2)
+}
+
+# Okay, we do have to do something, so parse matrix properly
+
+suppressPackageStartupMessages(require(DropletUtils))
+sce <- read10xCounts(opt$expression_data)
+
+metadata <- read.csv(opt$metadata, sep = "\t", stringsAsFactors = FALSE)
+
+for (field in c(opt$cell_id_field, opt$cell_type_field)){
+  if (! opt$cell_id_field %in% colnames(metadata)){
```

Comment:
it should be `field` here, I guess, not `opt$cell_id_field`

Reply:
yep!